File: cleanup-gateway/internal/server/control.go
package server

// The Suno Control route: POST /control plans exactly ONE action from the
// current screen. The sidecar proxies it byte-for-byte, so the wire contract
// here is the wire contract the Swift client sees:
//
//	POST /control  {"goal","steps":[{"action","note"}],"image","context"}
//	200            flat action JSON: {"action","x","y",...,"note","lease"}
//	400            {"error":"malformed json"|"malformed request","message":...}
//	402-class      from the account middleware (same entitlement as /cleanup)
//	429            {"error":"limit"} + Retry-After (60=daily/hard, 1=per-minute)
//	502            {"error":"unavailable"} — backend failure or unparseable plan
//
// It is a NON-streaming one-shot (unlike /answer): the client loop calls it
// once per step. No raw fallback ever — if the planner cannot answer with a
// valid action the loop stops; moving the user's cursor on a guess is the one
// thing this endpoint must never do.

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"image"
	_ "image/jpeg" // register JPEG for image.DecodeConfig (the screen image)
	_ "image/png"  // register PNG too, for parity with test fixtures
	"net/http"
	"strings"
	"time"

	"github.com/sunoflow/cleanup-gateway/internal/analytics"
	"github.com/sunoflow/cleanup-gateway/internal/backend"
	"github.com/sunoflow/cleanup-gateway/internal/caller"
	"github.com/sunoflow/cleanup-gateway/internal/cleanup"
	"github.com/sunoflow/cleanup-gateway/internal/control"
)

// maxControlBody caps the /control request body at 4MB — same budget as
// /answer: a 1600px JPEG screenshot (base64) plus prompt text with headroom.
const maxControlBody = 4 << 20

// controlRequest mirrors the /control API contract.
type controlRequest struct {
	// Goal is the user's dictated goal for this run.
	Goal string `json:"goal"`
	// Steps are the actions already taken this run, oldest first. Empty on
	// step one.
	Steps []control.Step `json:"steps"`
	// ImageB64 is the current screen, base64 JPEG. Empty means "no image" —
	// legal (the planner still answers, e.g. failed), though a run should
	// always carry one.
	ImageB64 string `json:"image"`
	// Dictionary is the user's own terms relevant to the goal, selected by
	// the sidecar from its local corrections file.
	Dictionary []cleanup.Entry `json:"dictionary"`
	// Context is what the client observed: frontmost app, focused window
	// title, cursor position in the screenshot's pixel space, and the
	// screenshot's dimensions (the client converts action coordinates to
	// screen points with them).
	Context controlContext `json:"context"`
}

// controlContext is the observed context block of a /control request.
type controlContext struct {
	App    string `json:"app"`
	Window string `json:"window"`
	// OS is the host operating system as the client reported it (name,
	// version, build, architecture). It pins the platform conventions the
	// planner must use; empty when the client is older and does not send it.
	OS          string `json:"os"`
	CursorX     *int   `json:"cursor_x"`
	CursorY     *int   `json:"cursor_y"`
	ImageWidth  *int   `json:"image_width"`
	ImageHeight *int   `json:"image_height"`
}

// controlResponse is the 200 body: the action fields flat, plus the lease.
type controlResponse struct {
	Action     string   `json:"action"`
	X          *int     `json:"x,omitempty"`
	Y          *int     `json:"y,omitempty"`
	X2         *int     `json:"x2,omitempty"`
	Y2         *int     `json:"y2,omitempty"`
	Text       *string  `json:"text,omitempty"`
	Key        *string  `json:"key,omitempty"`
	Modifiers  []string `json:"modifiers,omitempty"`
	Direction  *string  `json:"direction,omitempty"`
	Amount     *int     `json:"amount,omitempty"`
	Seconds    *float64 `json:"seconds,omitempty"`
	Note       *string  `json:"note,omitempty"`
	PressEnter *bool    `json:"press_enter,omitempty"`
	Lease      string   `json:"lease,omitempty"`
	// Usage is this step's token accounting, so the client can log per-step
	// cost. Omitted when the provider reported nothing (all zero).
	Usage *usageResponse `json:"usage,omitempty"`
}

// usageResponse is the per-step token block echoed to the client (for the
// control log). Mirrors backend.Usage; thinking tokens are billed as output
// but kept separate for visibility.
type usageResponse struct {
	PromptTokens   int `json:"prompt_tokens"`
	OutputTokens   int `json:"output_tokens"`
	ThinkingTokens int `json:"thinking_tokens"`
	TotalTokens    int `json:"total_tokens"`
}

// controlImageDims returns the screen image's pixel size for normalized→pixel
// conversion: the client-reported dimensions when present and sane, otherwise
// the image's own header, otherwise (0,0) — in which case the caller skips
// conversion rather than emit a wrong coordinate.
func controlImageDims(ctx controlContext, img []byte) (int, int) {
	if ctx.ImageWidth != nil && ctx.ImageHeight != nil && *ctx.ImageWidth > 0 && *ctx.ImageHeight > 0 {
		return *ctx.ImageWidth, *ctx.ImageHeight
	}
	if len(img) > 0 {
		if cfg, _, err := image.DecodeConfig(bytes.NewReader(img)); err == nil {
			return cfg.Width, cfg.Height
		}
	}
	return 0, 0
}

// controlBackend returns the control seam, or nil when the configured backend
// cannot plan actions (then /control answers 501).
func (s *Server) controlBackend() backend.ControlBackend {
	if cb, ok := s.Backend.(backend.ControlBackend); ok {
		return cb
	}
	return nil
}

// handleControl is the Suno Control endpoint. Authentication and the control
// quota run in the middleware chain; everything here is the one-shot plan.
func (s *Server) handleControl(w http.ResponseWriter, r *http.Request) {
	started := time.Now()

	cb := s.controlBackend()
	if cb == nil {
		writeJSON(w, http.StatusNotImplemented, map[string]string{"error": "unavailable", "message": "no control backend"})
		return
	}

	// Read the body under the per-route cap.
	var req controlRequest
	if err := json.NewDecoder(http.MaxBytesReader(w, r.Body, maxControlBody)).Decode(&req); err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "malformed json"})
		return
	}
	goal := strings.TrimSpace(req.Goal)
	if goal == "" {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "malformed request", "message": "empty goal"})
		return
	}

	// The screen image: like /answer, an absent one is legal (a text-only
	// turn — the planner may still answer failed), anything present must be
	// base64 JPEG under the cap.
	image, err := decodeImage(req.ImageB64)
	if err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "malformed request", "message": err.Error()})
		return
	}

	dict := cleanup.NormalizeDict(req.Dictionary)
	if len(dict) > control.MaxDictEntries {
		dict = dict[:control.MaxDictEntries]
	}

	// Which coordinate dialect the configured model speaks. Tool mode (native
	// computer_use) always answers 0-999, so it implies normalized; otherwise
	// it follows CONTROL_COORD_SPACE. Normalized coords are converted back to
	// image pixels below, so the client only ever sees pixels.
	normalized := s.ControlUseTool || strings.EqualFold(s.ControlCoordSpace, "normalized")

	// Build the prompt: framing → [SCREEN] → [STEPS ALREADY TAKEN] →
	// [DICTIONARY] → [GOAL] → context line. Tool mode drops the JSON schema and
	// coordinate rule (the tool defines both).
	steps := req.Steps
	if len(steps) > control.MaxSteps {
		steps = steps[len(steps)-control.MaxSteps:]
	}
	promptInput := control.Prompt{
		Goal:        goal,
		Steps:       steps,
		Dictionary:  dict,
		App:         req.Context.App,
		Window:      req.Context.Window,
		OS:          req.Context.OS,
		CursorX:     req.Context.CursorX,
		CursorY:     req.Context.CursorY,
		ImageWidth:  req.Context.ImageWidth,
		ImageHeight: req.Context.ImageHeight,
		Image:       len(image) > 0,
		Normalized:  normalized,
	}
	buildPrompt := control.BuildPrompt
	if s.ControlUseTool {
		buildPrompt = control.BuildToolPrompt
	}
	promptText := strings.Join(buildPrompt(promptInput), "\n")

	text, usage, err := cb.PlanAction(r.Context(), promptText, image)
	if err != nil {
		// A safety block is not an outage: the filter declined this specific
		// goal, so the client should say so (and the user can rephrase)
		// instead of "try again shortly". Report a distinct error code with a
		// fixed message; never echo prompt/goal content back.
		if errors.Is(err, backend.ErrSafetyBlock) {
			s.Logger.Warn("control plan blocked by safety filter")
			writeJSON(w, http.StatusBadGateway, map[string]string{
				"error":   "safety_block",
				"message": "Suno Control's AI planner declined this goal. Try phrasing it differently.",
			})
			return
		}
		s.Logger.Warn("control plan failed", "err", err.Error())
		code := http.StatusBadGateway
		body := map[string]string{"error": "unavailable"}
		if errors.Is(err, context.DeadlineExceeded) {
			body["message"] = "timed out"
		}
		writeJSON(w, code, body)
		return
	}

	action, err := control.ParseAction(text)
	if err != nil {
		// The planner spoke, but not in our vocabulary. This is a backend
		// failure for the loop's purposes: 502, no action, the client stops.
		s.Logger.Warn("control plan unparseable", "err", err.Error())
		writeJSON(w, http.StatusBadGateway, map[string]string{"error": "unavailable"})
		return
	}

	// Normalized models answer 0-1000 per axis; convert to image pixels so the
	// response carries the same pixel coordinates the pixel dialect does — the
	// client's executor is identical either way. Needs the image's size: the
	// client reports it, and the image header is the fallback.
	if normalized {
		if w, h := controlImageDims(req.Context, image); w > 0 && h > 0 {
			action.Denormalize(w, h)
		}
	}

	// Echo token usage to the client (control log) only when the provider
	// actually reported it, so omitempty drops the block otherwise.
	var usageResp *usageResponse
	if usage.TotalTokens > 0 || usage.PromptTokens > 0 {
		usageResp = &usageResponse{
			PromptTokens:   usage.PromptTokens,
			OutputTokens:   usage.OutputTokens,
			ThinkingTokens: usage.ThinkingTokens,
			TotalTokens:    usage.TotalTokens,
		}
	}

	writeJSON(w, http.StatusOK, controlResponse{
		Action:     action.Name,
		X:          action.X,
		Y:          action.Y,
		X2:         action.X2,
		Y2:         action.Y2,
		Text:       action.Text,
		Key:        action.Key,
		Modifiers:  action.Modifiers,
		Direction:  action.Direction,
		Amount:     action.Amount,
		Seconds:    action.Seconds,
		Note:       action.Note,
		PressEnter: action.PressEnter,
		Lease:      leaseFor(r),
		Usage:      usageResp,
	})

	// Analytics: zero content — lengths, counts, flags, timings, outcome.
	clientOS, clientVersion := parseClient(r.Header.Get(clientHeader))
	id, _ := caller.From(r.Context())
	props := map[string]any{
		"os":               clientOS,
		"app_version":      clientVersion,
		"had_image":        len(image) > 0,
		"step_index":       len(req.Steps),
		"goal_length":      len(goal),
		"action":           action.Name,
		"model":            s.ControlModel,
		"tool":             s.ControlUseTool,
		"latency_total_ms": time.Since(started).Milliseconds(),
		// Per-step token usage from the provider (zero when unreported), so cost
		// per step is measured, not estimated. Thinking tokens are billed as
		// output but reported apart from the visible answer.
		"prompt_tokens":   usage.PromptTokens,
		"output_tokens":   usage.OutputTokens,
		"thinking_tokens": usage.ThinkingTokens,
		"total_tokens":    usage.TotalTokens,
	}
	s.Analytics.Capture(analytics.Event{
		Name:       "control",
		DistinctID: id.MeterKey(),
		Properties: props,
		PersonProperties: map[string]any{
			"os":          clientOS,
			"app_version": clientVersion,
		},
	})

	// Also to the gateway log, so token usage per step is greppable during
	// calibration without opening the analytics dashboard.
	s.Logger.Info("control step",
		"model", s.ControlModel,
		"action", action.Name,
		"prompt_tokens", usage.PromptTokens,
		"output_tokens", usage.OutputTokens,
		"thinking_tokens", usage.ThinkingTokens,
		"total_tokens", usage.TotalTokens,
		"latency_ms", time.Since(started).Milliseconds())
}

File: sidecars/shared/control.py
"""Hosted Suno Control: proxy the gateway's one-action planner to the app.

Suno Control is the agent loop: the user speaks a goal, the app screenshots
the screen, and the gateway plans EXACTLY ONE action (click, type, key, ...)
that the app executes before asking again. Unlike /answer this is a one-shot
JSON call per step — no SSE. The sidecar is a dumb proxy: it applies the
user's dictation corrections to the goal (a mis-heard goal drives a mis-heard
plan), selects dictionary entries relevant to it (the same ``relevant_for``
machinery as /transcribe), and forwards. No prompt building, no fallback, no
retry: the gateway's response JSON passes through verbatim, and failures come
back as structured JSON the app can act on.
"""
import base64
import json
import os

import requests
from requests.adapters import HTTPAdapter, Retry

from sidecars.shared.cleanup import (
    CLIENT_ID,
    NotEntitled,
    _headers,
    _refusal,
)

CLEANUP_URL = os.environ.get("SUNOFLOW_CLEANUP_URL", "https://cleanup.ogcode.xyz/cleanup")
ENTITLEMENT_URL = CLEANUP_URL.rsplit("/", 1)[0] + "/entitlement"

# Mirrors cleanup.py's CLEANUP_URL: same host, /control path.
CONTROL_URL = os.environ.get("SUNOFLOW_CONTROL_URL", CLEANUP_URL.rsplit("/", 1)[0] + "/control")

# Dev-only key override, same posture as cleanup.py's CLEANUP_KEY: empty in
# production, the device key arrives per request from the app's Keychain.
CLEANUP_KEY = os.environ.get("SUNOFLOW_CLEANUP_KEY", "")

# Total ceiling on one planning step. The gateway's own deadline is 30s
# (CONTROL_TIMEOUT); the proxy waits slightly longer so the gateway's own
# error body — which says something kinder than a socket error would — is what
# arrives.
CONTROL_TIMEOUT = 35.0

# One pooled connection? No — a separate Session, deliberately (same posture
# as answer.py): a paid one-shot call must never queue behind or reuse cleanup's
# pooled socket, and its adapter carries NO retry (a retried POST /control
# would execute the same step twice when the first attempt reached the gateway
# but the connection died — worse than double-spend, because the second call's
# plan would execute on a screen the first call may already have changed).
_adapter = HTTPAdapter(max_retries=Retry(
    total=0, connect=0, read=0, status=0, backoff_factor=0,
))
_session = requests.Session()
_session.mount("https://", _adapter)
_session.mount("http://", _adapter)

# Caps mirroring the gateway's. The sidecar is the first line of defense: it
# drops rather than errors, so a pathological client cannot balloon the request.
MAX_GOAL_LEN = 2000
MAX_STEPS = 100
MAX_STEP_NOTE_LEN = 300
MAX_DICT_ENTRIES = 64
MAX_IMAGE_BYTES = 3 << 20

_NOT_CONNECTED = (
    "This device isn't connected to a SunoFlow account. "
    "Open Settings → Account to connect it."
)


def plan_action(goal, steps=None, image_bytes=None, context=None, key="", dictionary=None):
    """One planning step: POST to the gateway and return its JSON.

    Raises NotEntitled when the gateway refuses the device (the caller turns
    that into the same 402 response /transcribe returns). Raises ValueError on
    a bad goal/image from the app itself. On a limit, outage, or gateway
    502/4xx it returns {"error": ...} — the app's loop stops on anything that
    is not an action.
    """
    if not key:
        raise NotEntitled(_NOT_CONNECTED, code="not_connected")

    goal = (goal or "").strip()[:MAX_GOAL_LEN]
    if not goal:
        raise ValueError("empty goal")

    bounded = []
    for s in list(steps or []):
        if not isinstance(s, dict):
            continue
        action = str(s.get("action") or "").strip()[:32]
        if not action:
            continue
        bounded.append({"action": action, "note": str(s.get("note") or "").strip()[:MAX_STEP_NOTE_LEN]})
    bounded = bounded[-MAX_STEPS:]

    dictionary = list(dictionary or [])[:MAX_DICT_ENTRIES]

    payload = {
        "goal": goal,
        "steps": bounded,
        "context": context or {},
    }
    if dictionary:
        payload["dictionary"] = dictionary
    if image_bytes:
        if len(image_bytes) > MAX_IMAGE_BYTES:
            raise ValueError("image too large")
        payload["image"] = base64.b64encode(image_bytes).decode("ascii")

    # Deliberately NO retry (contrast cleanup's single-retry adapter) — see the
    # module docstring: a retried plan would double-execute on a changed screen.
    try:
        resp = _session.post(
            CONTROL_URL,
            headers=_headers(key),
            json=payload,
            timeout=(10, CONTROL_TIMEOUT),
        )
    except Exception:
        # Network-level failure before any byte: not an entitlement question —
        # the request may never have arrived. The loop stops cleanly.
        return {"error": "unavailable", "message": "Suno Control is unavailable right now. Try again shortly."}

    refusal, code = _refusal(resp)
    if refusal:
        raise NotEntitled(refusal, code=code or "not_entitled")

    if resp.status_code == 429:
        retry_after = resp.headers.get("Retry-After", "60")
        if retry_after.strip() == "1":
            return {
                "error": "limit",
                "message": "Suno Control is working as fast as it can — wait a few seconds and try again.",
                "retry_after": 1,
            }
        return {
            "error": "limit",
            "message": "You've used all your Suno Control for today. It resets tomorrow.",
            "retry_after": 60,
        }

    if resp.status_code == 502:
        # The gateway reached the planner and it failed to produce a valid
        # action: the loop stops rather than execute on a guess. A safety
        # block is distinct — the planner declined THIS goal, not the service —
        # so its code and message pass through verbatim and the app can say
        # "rephrase" instead of "outage".
        try:
            body = resp.json()
        except ValueError:
            body = {}
        if isinstance(body, dict) and body.get("error") == "safety_block":
            return {
                "error": "safety_block",
                "message": str(body.get("message") or "Suno Control's AI planner declined this goal. Try phrasing it differently."),
            }
        return {"error": "unavailable", "message": "Suno Control couldn't decide the next step. Try again shortly."}

    if not resp.ok:
        try:
            body = resp.json()
        except ValueError:
            body = {}
        if isinstance(body, dict) and body.get("error") == "safety_block":
            return {
                "error": "safety_block",
                "message": str(body.get("message") or "Suno Control's AI planner declined this goal. Try phrasing it differently."),
            }
        return {"error": "unavailable", "message": "Suno Control is unavailable right now. Try again shortly."}

    try:
        action = resp.json()
    except ValueError:
        return {"error": "unavailable", "message": "Suno Control is unavailable right now. Try again shortly."}

    if not isinstance(action, dict) or not action.get("action"):
        return {"error": "unavailable", "message": "Suno Control is unavailable right now. Try again shortly."}

    return action